### ISSUE TITLE
WebSocket and HTTP on the same API Definition

### ISSUE THREAD

**Issue by dikderoy** [TykTechnologies/tyk#1771]
**Do you want to request a *feature* or report a *bug*?**
BUG
**What is the current behavior?**
It is impossible to use websocket and plain-HTTP on the same endpoint (or two endpoints with shared path prefix) defined in single API Definition

**What is the expected behavior?**
Full support of WebSocket and plain-HTTP on per-request/per-protocol basis (I understand what you are caching upstream connections, but this leads to described problem)

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem**

API definition is set to forward any traffic under `/path/{rest:*}` to `http://service_endpoint:8080/`
and for example:
- `/path/socket` is WS endpoint
- `/path/request` is plain HTTP endpoint

then after establishing WS connection on `/path/socket` plain-http requests to `/path/request` will fail (no response).

Following requests to websocket endpoint with WS protocol also fail **until Tyk Gateway is restarted.**

Following appears in log:

```
tyk_gateway_1      | 2018/06/09 19:49:50 http: panic serving 172.19.0.21:54130: interface conversion: http.RoundTripper is *main.WSDialer, not *http.Transport
tyk_gateway_1      | goroutine 3655 [running]:
tyk_gateway_1      | net/http.(*conn).serve.func1(0xc4201f3040)
tyk_gateway_1      | 	/usr/local/go/src/net/http/server.go:1697 +0xd0
tyk_gateway_1      | panic(0xd7d8c0, 0xc4206b0640)
tyk_gateway_1      | 	/usr/local/go/src/runtime/panic.go:491 +0x283
tyk_gateway_1      | main.(*ReverseProxy).WrappedServeHTTP(0xc4202f5d60, 0x14cee40, 0xc42043b500, 0xc4205c7200, 0x28881600, 0x0)
tyk_gateway_1      | 	/src/github.com/TykTechnologies/tyk/reverse_proxy.go:591 +0x2008
tyk_gateway_1      | main.(*ReverseProxy).ServeHTTP(0xc4202f5d60, 0x14cee40, 0xc42043b500, 0xc4205c7200, 0x9)
tyk_gateway_1      | 	/src/github.com/TykTechnologies/tyk/reverse_proxy.go:362 +0x60
tyk_gateway_1      | main.(*SuccessHandler).ServeHTTP(0xc4201b1f00, 0x14cee40, 0xc42043b500, 0xc4205c7200, 0xffffffffffffffff)
tyk_gateway_1      | 	/src/github.com/TykTechnologies/tyk/handler_success.go:245 +0x167
tyk_gateway_1      | main.(*DummyProxyHandler).ServeHTTP(0xc4201b1f00, 0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/src/github.com/TykTechnologies/tyk/api_loader.go:521 +0x4b
tyk_gateway_1      | main.createMiddleware.func1.1(0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/src/github.com/TykTechnologies/tyk/middleware.go:101 +0x76d
tyk_gateway_1      | net/http.HandlerFunc.ServeHTTP(0xc4205d6600, 0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/usr/local/go/src/net/http/server.go:1918 +0x44
tyk_gateway_1      | main.createMiddleware.func1.1(0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/src/github.com/TykTechnologies/tyk/middleware.go:101 +0x76d
tyk_gateway_1      | net/http.HandlerFunc.ServeHTTP(0xc4205d6640, 0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/usr/local/go/src/net/http/server.go:1918 +0x44
tyk_gateway_1      | main.createMiddleware.func1.1(0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/src/github.com/TykTechnologies/tyk/middleware.go:101 +0x76d
tyk_gateway_1      | net/http.HandlerFunc.ServeHTTP(0xc4205d6680, 0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/usr/local/go/src/net/http/server.go:1918 +0x44
tyk_gateway_1      | main.createMiddleware.func1.1(0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/src/github.com/TykTechnologies/tyk/middleware.go:101 +0x76d
tyk_gateway_1      | net/http.HandlerFunc.ServeHTTP(0xc4205d66c0, 0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/usr/local/go/src/net/http/server.go:1918 +0x44
tyk_gateway_1      | main.createMiddleware.func1.1(0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/src/github.com/TykTechnologies/tyk/middleware.go:101 +0x76d
tyk_gateway_1      | net/http.HandlerFunc.ServeHTTP(0xc4205d6700, 0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/usr/local/go/src/net/http/server.go:1918 +0x44
tyk_gateway_1      | main.createMiddleware.func1.1(0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/src/github.com/TykTechnologies/tyk/middleware.go:101 +0x76d
tyk_gateway_1      | net/http.HandlerFunc.ServeHTTP(0xc4205d6740, 0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/usr/local/go/src/net/http/server.go:1918 +0x44
tyk_gateway_1      | main.createMiddleware.func1.1(0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/src/github.com/TykTechnologies/tyk/middleware.go:101 +0x76d
tyk_gateway_1      | net/http.HandlerFunc.ServeHTTP(0xc4205d6780, 0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/usr/local/go/src/net/http/server.go:1918 +0x44
tyk_gateway_1      | main.createMiddleware.func1.1(0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/src/github.com/TykTechnologies/tyk/middleware.go:101 +0x76d
tyk_gateway_1      | net/http.HandlerFunc.ServeHTTP(0xc4205d67c0, 0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/usr/local/go/src/net/http/server.go:1918 +0x44
tyk_gateway_1      | main.createMiddleware.func1.1(0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/src/github.com/TykTechnologies/tyk/middleware.go:101 +0x76d
tyk_gateway_1      | net/http.HandlerFunc.ServeHTTP(0xc4205d6800, 0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/usr/local/go/src/net/http/server.go:1918 +0x44
tyk_gateway_1      | github.com/TykTechnologies/tyk/vendor/github.com/gorilla/mux.(*Router).ServeHTTP(0xc420058c60, 0x14cee40, 0xc42043b500, 0xc4205c7200)
tyk_gateway_1      | 	/src/github.com/TykTechnologies/tyk/vendor/github.com/gorilla/mux/mux.go:150 +0xed
tyk_gateway_1      | main.mainHandler.ServeHTTP(0x14cee40, 0xc42043b500, 0xc4205c7000)
tyk_gateway_1      | 	/src/github.com/TykTechnologies/tyk/main.go:1161 +0x73
tyk_gateway_1      | main.(*mainHandler).ServeHTTP(0x15d6f80, 0x14cee40, 0xc42043b500, 0xc4205c7000)
tyk_gateway_1      | 	<autogenerated>:1 +0x52
tyk_gateway_1      | net/http.serverHandler.ServeHTTP(0xc420070ea0, 0x14cee40, 0xc42043b500, 0xc4205c7000)
tyk_gateway_1      | 	/usr/local/go/src/net/http/server.go:2619 +0xb4
tyk_gateway_1      | net/http.(*conn).serve(0xc4201f3040, 0x14cfe40, 0xc4207164c0)
tyk_gateway_1      | 	/usr/local/go/src/net/http/server.go:1801 +0x71d
tyk_gateway_1      | created by net/http.(*Server).Serve
tyk_gateway_1      | 	/usr/local/go/src/net/http/server.go:2720 +0x288
```

**Which versions of Tyk affected by this issue? Did this work in previous versions of Tyk?**
Tyk Gateway 2.6.1 and older

**Comment by glfp** [comment]
+1
I have the same issue when using Firefox. Firefox WS call pass firstly thru a proxy (that act as authenticator), then call Tyk. Due a invalid SSL configuration of the authenticator (it use self-signed certificate), Firefox WS call fall in error and automatically try to call the same endpoint with https: so the WS endpoint in Tyk is called in https and crash.

This behavior occurs ever.

**Comment by dikderoy** [comment]
I would've considered this a CRITICAL and a DoS vulnerability for WS-endpoints - cause any attempt to perform HTTP connection on WS endpoint on TykGateway is effectively renders endpoint DEAD. no other WS connections could be established after doing this until TykGateway process is restarted. At least fix this "dead until restart" problem please.

**Comment by buger** [comment]
@dencoded pls take a look

**Comment by dencoded** [comment]
ok, I think I know what is going on, working on this at the moment

**Comment by glfp** [comment]
Hello, any news about this bug ? I'm going in production next week but I need to know if I need to block Firefox clients or not.

Thansk !

**Comment by dencoded** [comment]
@glfp PR is submitted for review just now, you should see this fix in next patch release (if everything goes well and it gets merged)

**Comment by dikderoy** [comment]
Any updates?

**Comment by buger** [comment]
Will be included in 2.7.2 with eta 2-3 weeks.
